server/src/main/java/com/thoughtworks/go/server/service/StageManualTriggerChecker.java

```diff
@@ -0,0 +1,73 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.service;
+
+import com.thoughtworks.go.config.Approval;
+import com.thoughtworks.go.config.CaseInsensitiveString;
+import com.thoughtworks.go.config.StageConfig;
+import com.thoughtworks.go.domain.Pipeline;
+import com.thoughtworks.go.domain.Stage;
+import com.thoughtworks.go.domain.StageResult;
+import com.thoughtworks.go.server.service.result.OperationResult;
+import com.thoughtworks.go.serverhealth.HealthStateScope;
+import com.thoughtworks.go.serverhealth.HealthStateType;
+
+public class StageManualTriggerChecker implements SchedulingChecker {
+
+    private final String pipelineName;
+    private final Integer pipelineCounter;
+    private final String stageName;
+    private final GoConfigService goConfigService;
+    private final PipelineService pipelineService;
+
+    public StageManualTriggerChecker(String pipelineName, Integer pipelineCounter, String stageName, GoConfigService goConfigService, PipelineService pipelineService) {
+        this.pipelineName = pipelineName;
+        this.pipelineCounter = pipelineCounter;
+        this.stageName = stageName;
+        this.goConfigService = goConfigService;
+        this.pipelineService = pipelineService;
+    }
+
+    @Override
+    public void check(OperationResult result) {
+        HealthStateType healthStateType = HealthStateType.general(HealthStateScope.forPipeline(pipelineName));
+        Pipeline pipeline = pipelineService.fullPipelineByCounter(pipelineName, pipelineCounter);
+        if (pipeline == null) {
+            result.success(healthStateType);
+            return;
+        }
+
+        if (!pipeline.hasStageBeenRun(stageName)) {
+            if (goConfigService.hasPreviousStage(pipelineName, stageName)) {
+                CaseInsensitiveString previousStageName = goConfigService.previousStage(pipelineName, stageName).name();
+                StageConfig currentStageConfig = goConfigService.nextStage(pipelineName, previousStageName.toString());
+                Approval approval = currentStageConfig.getApproval();
+                if (approval.isAllowOnlyOnSuccess()) {
+                    Stage previousStage = pipeline.getStages().byName(previousStageName.toString());
+                    StageResult previousStageResult = previousStage.getResult();
+                    if (previousStageResult != StageResult.Passed) {
+                        String errorMessage = String.format("Cannot schedule %s as the previous stage %s has %s!", stageName, previousStageName.toString(), previousStageResult.name());
+                        healthStateType = HealthStateType.forbidden();
+                        result.notAcceptable(errorMessage, healthStateType);
+                        return;
+                    }
+                }
+            }
+        }
```

Comment:
Some of this looks like duplicate code from `ScheduleService#checkIfPreviousStagePassed`, think of de-duplicating some of this?

Reply:
Done. Moved the logic to `SchedulingCheckerService.shouldAllowSchedulingStage`